### PR TITLE
cmd/snap-update-ns: re-factor secureMkdirAll into secureMk{Prefix,Dir}

### PR DESCRIPTION
**Pull request by zyga**
The act of "securely creating a directory" can be decomposed into
"securely create the prefix" and "securely create the leaf". Naturally
the leaf can now vary and we can easily create leaf directories, files
and symlinks.

This patch sets the stage for that to happen, by splitting the main
function into a pair of functions. On the outside nothing changes but
tests needed a slight tweak to accommodate how life-cycle of file
descriptors looks like.

There are also several new tests to improve code coverage in various
edge cases.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: cmd/snap-update-ns/utils.go [hunk 1]
```diff
@@ -45,6 +47,84 @@ var (
...
+		}
+	}
+
+	logger.Debugf("secure-mk-prefix %q %v %d %d -> %d", segments, perm, uid, gid, fd)
+	return fd, nil
+}
+
+// secureMkdir creates a directory at i-th entry of absolute path represented
+// by segments. This function can be used to construct subsequent elements of
+// the constructed path.
```
Comment (bboozzoo) [inline cmd/snap-update-ns/utils.go]: Maybe add a note that `secureMkDir` returns a file descriptor to newly created segment and the it's -1 on errors.

File: cmd/snap-update-ns/utils.go [hunk 2]
```diff
@@ -58,53 +138,29 @@ var (
...
 
 	// Only support absolute paths to avoid bugs in snap-confine when
 	// called from anywhere.
 	if !filepath.IsAbs(name) {
 		return fmt.Errorf("cannot create directory with relative path: %q", name)
 	}
-	// Open the root directory and start there.
-	fd, err = sysOpen("/", openFlags, 0)
+	segments := strings.FieldsFunc(filepath.Clean(name), func(c rune) bool { return c == '/' })
+
```
Comment (bboozzoo) [inline cmd/snap-update-ns/utils.go]: You should probably check if `len(segments) > 0`. Haven't tracked through the code but I think you'll get a panic with `name = "/"`.

File: cmd/snap-update-ns/utils_test.go [hunk 3]
```diff
@@ -53,20 +61,67 @@ func (s *utilsSuite) TestSecureMkdirAllRelative(c *C) {
...
-		`open "/" O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY 0`,
-		`mkdirat 3 "abs" 0755`,
-		`openat 3 "abs" O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY 0`,
+		`open "/" O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY 0`, // -> 3
+		`close 3`,
+	})
+}
+
+// Ensure that we can create a directory in the top-level directory.
+func (s *utilsSuite) TestSecureMkdirAllLevel1(c *C) {
```
Comment (pedronis) [inline cmd/snap-update-ns/utils_test.go]: don't we need a test for for update.SecureMkDirAll("/") ? should that fail if perm or gid/uid aren't the only possible ones?

File: cmd/snap-update-ns/utils.go [hunk 4]
```diff
@@ -45,6 +47,87 @@ var (
...
+// by segments. This function can be used to construct subsequent elements of
+// the constructed path. The return value contains the newly created file
+// descriptor or -1 on error.
+func secureMkDir(fd int, segments []string, i int, perm os.FileMode, uid, gid int) (int, error) {
+	logger.Debugf("secure-mk-dir %d %q %d %v %d %d -> ...", fd, segments, i, perm, uid, gid)
+
+	segment := segments[i]
+	made := true
+	var err error
+	var newFd int
```
Comment (stolowski) [inline cmd/snap-update-ns/utils.go]: I'd init newFd to -1 to be on the safe side.

File: cmd/snap-update-ns/utils.go [hunk 5]
```diff
@@ -45,6 +47,87 @@ var (
...
+// Create directories for all but the last segments and return the file
+// descriptor to the leaf directory. This function is a base for secure
+// variants of mkdir, touch and symlink.
+func secureMkPrefix(segments []string, perm os.FileMode, uid, gid int) (int, error) {
+	logger.Debugf("secure-mk-prefix %q %v %d %d -> ...", segments, perm, uid, gid)
+
+	// Declare var and don't assign-declare below to ensure we don't swallow
+	// any errors by mistake.
+	var err error
+	var fd int
```
Comment (stolowski) [inline cmd/snap-update-ns/utils.go]: I'd init fd to -1 to be on the safe side.

Comment (zyga) [reply cmd/snap-update-ns/utils.go]: Commented in the other PR that we cannot do that as it triggers ineffectual assignment error.

File: cmd/snap-update-ns/utils_test.go (hunk 3 above)
Comment (zyga) [reply cmd/snap-update-ns/utils_test.go]: I don't think this needs to be this paranoid. First of all, we don't chown things we don't create and we don't create / ever :-) (we just open it, there's no code to try to create it). Lastly we are still confined and we cannot open / in writable mode.

Comment (pedronis) [reply cmd/snap-update-ns/utils_test.go]: either way we still need to test what happens

Comment (zyga) [reply cmd/snap-update-ns/utils_test.go]: I added a Level0 test here actually, let me see ...

File: cmd/snap-update-ns/utils.go [hunk 6]
```diff
@@ -45,6 +47,87 @@ var (
 	sysFchown  = syscall.Fchown
 )
 
+// Create directories for all but the last segments and return the file
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: s/segments/segment/

File: cmd/snap-update-ns/utils.go [hunk 7]
```diff
@@ -45,6 +47,87 @@ var (
...
+// descriptor to the leaf directory. This function is a base for secure
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: Rather than 'leaf directory' I think you mean 'parent'. Honestly, I found the naming here a bit confusing. secureMkDirParent() might be more clear...

I still find it code hard to read and the decomposition feels awkward where secureMkDirPrefix is passed all the segments (elements or tokens might be a better name for the variable...) and it creates everything but the last one. Why not just pass secureMkDirAll the parent and be done with it?

Put another way, why 3 functions: secureMkDirAll which creates everything, secureMkDirPrefix which creates all but the last and secureMkDir which creates an arbitrary internal element? Why isn't it sufficient to have a single function that just goes from '/' to the path you want? I guess I don't understand the point of this PR-- can you describe where you are going and why this needs to be decomposed in this way?

File: cmd/snap-update-ns/utils.go [hunk 8]
```diff
@@ -45,6 +47,87 @@ var (
...
+			return -1, fmt.Errorf("cannot mkdir path segment %q: %v", segment, err)
+		}
+	}
+	newFd, err = sysOpenat(fd, segment, openFlags, 0)
+	if err != nil {
+		return -1, fmt.Errorf("cannot open path segment %q (got up to %q): %v", segment,
+			"/"+strings.Join(segments[:i], "/"), err)
+	}
+	if made {
+		// Chown each segment that we made.
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: This comment is misleading. You only created one segment. s/each/the/

File: cmd/snap-update-ns/utils.go [hunk 9]
```diff
@@ -58,53 +141,31 @@ var (
...
-			case syscall.EEXIST:
-				made = false
-			default:
-				return fmt.Errorf("cannot mkdir path segment %q: %v", segment, err)
-			}
-		}
-		fd, err = sysOpenat(fd, segment, openFlags, 0)
+	if len(segments) > 0 {
+		// Create the final segment as a directory.
+		fd, err = secureMkDir(fd, segments, len(segments)-1, perm, uid, gid)
```
Comment (jdstrand) [inline cmd/snap-update-ns/utils.go]: Again, this is awkward without know the reason why you broke it up in this way. We do:
```
segments := ...
...
fd, err := secureMkPrefix(segments, perm, uid, gid)
...
fd, err = secureMkDir(fd, segments, len(segments)-1, perm, uid, gid)
```

To me, an API something along of this seems more clear:
```
// create the dir and any needed parent directories
createParents := true
fd, err := secureMkDir(segments, perm, uid, gid, createParents)
...
// create only the specified directory and no parents
createParents = false
fd, err := secureMkDir(segments, perm, uid, gid, createParents)
```

Then if you just want to create the parent directory (and its parents) for some reason, you just do:
```
fd, err := secureMkDir(segments[0:len(segments)-1], perm, uid, gid, true)
```

**Comment by jdstrand** [comment]
Unfortunately this was merged without my review (I said I would complete it by today but by the time I got to it, it was merged).

File: cmd/snap-update-ns/utils.go (hunk 7 above)
Comment (pedronis) [reply cmd/snap-update-ns/utils.go]: afaict the description says that the plan is to reuse and compose *Prefix together with functions that create symlinks or files instead of dirs

Comment (jdstrand) [reply cmd/snap-update-ns/utils.go]: That doesn't explain why the API is decomposed and implemented in the manner it is-- maybe its just me, but, while it works, I found it somewhat convoluted.

Comment (jdstrand) [reply cmd/snap-update-ns/utils.go]: For example, having one function to create a directory, one to create a symlink and one to create a file would be clear. The symlink and the file functions would simply call the directory function with the dirname of the symlink/file to make sure the directory was created.
